distributed/service.py

```diff
@@ -0,0 +1,426 @@
+        to the data.
+
+        If ``produce_key`` raises an error, the state of the data is undefined. In most
+        cases, the `Service` should call `restart` if this happens.
+        """
+        ...
+
+    async def restart(self) -> None:
+        """
+        Request that the service task be restarted.
+
+        `Service.stop` is guaranteed not to be called until after `restart` has
+        returned.
+
+        Once `restart` has returned, `Service.stop` will be called on all instances for
+        the service task, and workers will release their references those instances. Once
+        `stop` has returned on all instances, the task will transition to ``waiting`` on
+        the scheduler. Then, it will transition back to ``processing``, and new
+        `Service` instances (with the same service task key) will be started on all
+        workers.
+        """
+        ...
+
+    async def error(self, exc: Exception) -> None:
+        """
+        Fail the service task with an error message.
+
+        `Service.stop` is guaranteed not to be called until after `error` has returned.
+
+        Once `error` has returned, `Service.stop` will be called on all instances for
+        the service task, and workers will release their references those instances. Once
+        `stop` has returned on all instances, the task will transition to ``erred`` on
+        the scheduler. The result of the task will be ``exc``.
+
+        If `error` is called multiple times, only the first ``exc`` to reach the
+        scheduler will become the result; all others will be dropped (though all calls
+        to `error` are logged on the `Worker`).
+        """
+        ...
+
+    async def run_in_executor(
+        self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs
+    ) -> Awaitable[R]:
+        """
+        Schedule a function to run in the Worker's default executor.
+
+        This will take up a slot in `Worker.executing_count`, reducing the number of
+        other tasks (or `run_in_executor` calls) that can run concurrently. This allows
+        multiple Services (or Services and normal tasks) to cooperatively run
+        resource-intensive synchronous functions concurrently without oversubscribing
+        the `Worker`. This way, `Worker.nthreads` is respected even for out-of-band
+        functions.
+
+        For event-loop-blocking functions that aren't resource-intensive (network calls,
+        etc.), `distributed.compatibility.to_thread` is preferred.
+        """
+        ...
+
+
+# TODO: how to type annotate RPCs? https://github.com/python/mypy/issues/5523
+class ServiceHandle(Generic[T]):
+    "RPC to a Service instance running on another worker"
+
+    def __getattr__(self, attr: str) -> _RPC:
+        ...
+
+
+# https://stackoverflow.com/a/65564936/17100540
+class _RPC(Protocol):
+    def __call__(self, *args, **kwargs) -> Awaitable:
+        ...
```

Comment:
I don't think we should reinvent a new RPC framework but instead improve on the existing one. I don't see a reason why they should be different, is there?